apis/ninedof/src/lib.rs

```diff
@@ -0,0 +1,211 @@
+#![no_std]
+
+use core::cell::Cell;
+use libtock_platform::{
+    share, share::Handle, subscribe::OneId, DefaultConfig, ErrorCode, Subscribe, Syscalls, Upcall,
+};
+
+pub struct NineDof<S: Syscalls>(S);
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct NineDofData {
+    pub x: i32,
+    pub y: i32,
+    pub z: i32,
+}
+
+fn sqrt(n: f64) -> f64 {
+    /// Newton-Raphson method for sqrt
+    const EPSILON: f64 = 0.00001;
+    let mut x: f64 = n;
+    let mut y: f64 = 1.0;
+
+    while x - y > EPSILON {
+        x = (x + y) / 2.0;
+        y = n / x;
+    }
+
+    x
+}
+
+impl<S: Syscalls> NineDof<S> {
+    /// Returns Ok() if the driver was present.This does not necessarily mean
+    /// that the driver is working.
+    pub fn exists() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, EXISTS, 0, 0).to_result()
+    }
+
+    /// Initiate a accelerometer measurement.
+    /// This function is used both for synchronous and asynchronous readings
+    pub fn read_accelerometer() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, READ_ACCELEROMETER, 0, 0).to_result()
+    }
+
+    /// Initiate a magnetometer measurement.
+    /// This function is used both for synchronous and asynchronous readings
+    pub fn read_magnetometer() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, READ_MAGNETOMETER, 0, 0).to_result()
+    }
+
+    /// Initiate a gyroscope measurement.
+    /// This function is used both for synchronous and asynchronous readings
+    pub fn read_gyro() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, READ_GYRO, 0, 0).to_result()
+    }
+
+    /// Unregister an events listener
+    pub fn unregister_listener() {
+        S::unsubscribe(DRIVER_NUM, 0)
+    }
+
+    /// Register an events listener
+    pub fn register_listener<'share, F: Fn(NineDofData)>(
+        listener: &'share NineDofListener<F>,
+        subscribe: Handle<Subscribe<'share, S, DRIVER_NUM, 0>>,
+    ) -> Result<(), ErrorCode> {
+        S::subscribe::<_, _, DefaultConfig, DRIVER_NUM, 0>(subscribe, listener)
+    }
+
+    /// Initiate a synchronous accelerometer measurement.
+    /// Returns Ok(accelerometer_value) if the operation was successful
+    /// Returns Err(ErrorCode) if the operation was unsuccessful
+    pub fn read_accelerometer_sync(x: &mut i32, y: &mut i32, z: &mut i32) -> Result<(), ErrorCode> {
```

Comment:
```suggestion
    pub fn read_accelerometer_sync() -> Result<NineDofData, ErrorCode> {
```

Reply:
I have modified those and don't apper in the last commit, they're outdated